docs/architecture/adr-007.md

```diff
@@ -0,0 +1,96 @@
+# ADR 007: Generic event formats for evm-hook actions
+
+## Changelog
+* 9-06-2021: Initial Draft
+
+## Context
+
+The current evm-hooks relies on event emitted from a smart contract to trigger specific actions upon contract execution.
+The current way to define those events is not flexible: events are hardcoded, parameters fixed and correspond only to one specific action.
+
+This can lead to potential issue in the long run when commands needs to be updated to add more parameters (ibc-go/ gravity bridge) or even if we need to implement more "actions". Since contracts are immutable, contract developers won't be able to have access to these functionalities after deploying their CRC20 contracts if we perform an upgrade.
+
+In this ADR, we define generic events format that can be use by smart contract developer to call any "action" upon successful contract execution. 
+
+## Decision
+
+We define 2 specifics events that smart contract can use to trigger specific action after contract execution.
+
+```
+# Trigger an action after contract execution which only involve the sender account
+__CronosCall(string functionName, address sender,  byte[]args)
+
+# Trigger an action after contract execution that require moving asset from the smart contract "account"
+__CronosCallAfterPayment(string functionName, address payer, uint amount, byte[]args)
+
+```
+
+- **functionName** corresponds to the equivalent hardcoded eventName such as `CronosSendToIBC` for the event `__CronosSendToIbc` or `CronosSendToChain` for the event `__CronosSendToChain`
+- **payer** is the address which make the payment, usually equal to `msg.sender`
+- **amount** corresponds to the `amount` that the payer has paid (either by burning or sending asset to the SC)
+- **args** corresponds to the argument in order of the event removing the `sender` and `amount` from the parameters encoded using the abi event pack
+
+
+
+An example of smart contract implementing those function
+
+```
+pragma solidity ^0.6.8;
+
+import "ds-token/token.sol";
+
+Contract Demo is DSToken {
+
+ __CronosCallAfterPayment(string functionName, address sender, uint amount, byte[]args)
+ 
+ // unsafe_burn burn tokens without user's approval and authentication, used internally
+ function unsafe_burn(address addr, uint amount) internal {
+     // Deduct user's balance without approval
+     require(balanceOf[addr] >= amount, "ds-token-insufficient-balance");
+     balanceOf[addr] = sub(balanceOf[addr], amount);
+     totalSupply = sub(totalSupply, amount);
+     emit Burn(addr, amount);
+ }
+ 
+ // send an "amount" of the contract token to recipient through IBC
+ // args should contains the encoded value recipient or extra value if needed in the future
+ function send_to_ibc(uint amount, byte[] args) public {
+     unsafe_burn(msg.sender, amount);
+     emit __CronosCallAfterPayment("CronosSendToIbc", msg.sender, amount, args);
```

Comment:
Will we recommend apps to allow user to pass in the function name as well?

Reply:
I guess it depend of the smart contract intent.

but the function name can also be an input parameter if the smart contract is a controller for ica for example and they expects more features to be implemented

Reply:
It's about convention and expectation, for example, we want to extend sth for ica, would we add a new function name, or just support extend the args format, in the first case, this demo contract still need to upgrade itself.

Reply:
yeah I guess it also depend on what kind of actions we need to implement in the future

I guess for now the focus is on "native token transfer capability" from smart contract account to a "recipient".

If we support both way (hardcoded event or generic events), the sc developers can either choose between exposing limited functions and define different payment rule for each transfer methods (but it takes the risk to not be compatible in the future) or have a more generic capability that does not requires contract upgrade

Reply:
what if we just add a some really generic stuff like this:

```
event __CronosGenericCall(address, uint, byte[])

function cronos_generic_call(uint amount, byte[] payload) public {
  if amount > 0 {
    unsafe_burn(msg.sender, amount);
  }
  emit __CronosGenericCall(msg.sender, amount, payload);
}
```

Reply:
you mean putting the function name in byte[]?

I guess this is fine, we need a better way to serialised the function name and input parameters available for smart contract devs

Reply:
and don't need to separate `__CronosCall` and `__CronosCallAfterPayment`.

Reply:
mmm the reason it is separated is for clarity. It is not clear what is the "amount" if you call name the function `__CronosCall` 

I also expect that having a generic function `__CronosCall` is not impossible at our stage and we might need to implement more "generic events" in the future for different type of actions

Reply:
the sematic is similar to calling a smart contract method with a value.

Reply:
anyway, it's a trade-off here, too generic may mean unexpected behavior in the future, and less generic means smart contract upgrade, personally, I prefer explicit contract upgrade to implicit stuff actually.
when an app wants to use a non-trivial new feature like gravity-bridge, it's reasonable to ask for an upgrade.
It's just the chain has gov module already, so we can do the upgrade relatively painlessly. but the smart contract can do govenance too, just ask the community to vote or sth.

Reply:
```
But the smart contract can do govenance too, allowing the community to vote or sth.
```

Do you mean bytecode replacement? if its case by case it seems hard to do... and also controversial

I guess our problem reside on the CRC20 contract that are not supposed to be upgradable.

To summarise from the less generic to the most generic solution:

1. Just make the current events "more flexible" for example `__CronosSendToIbc(address, uint, bytes[]arg)` to make sure they will still be compatible after upgrade

2. Make a generic function only for token related stuff ` __CronosCallAfterPayment(string functionName, address sender, uint amount, byte[]args)` which can be used to perform actions on contract's assets which is basically what this ADR propose

3. Try to define a unique generic event (at least best effort) that can be use by contract developers such as `__CronosGenericCall(address, uint, byte[])` or `__CronosGenericCall(byte[], byte[])` which allows to call any methods and future hooks from the smart contract

Reply:
> Do you mean bytecode replacement?

I mean the smart contract should do the upgrade for themselves if they want to bring in a significant new feature like gravity-bridge.
What we are proposing here is trying to smuggle new features into the smart contract through chain upgrade.

Reply:
```
What we are discussing here is trying to smuggle new features into the smart contract through chain upgrade.
```

Technically those event hooks are a way for user to chain a evm tx with a cosmos tx upon contract execution

I can see two big use cases

- Trigger an action on behalf msg.sender account. In that case, using the smart contract as a gateway is just a way for us to "expose" those functionalities to metamask (in the effort to be compatible with ethereum)

- Trigger an action on behalf of the smart contract account. Smart contracts have the ability to move assets (coin, nft etc) from their accounts and perform diverse actions (ibc, bridge etc)

In that case does is it okay to give those abilities to all the smart contracts in general? the only issue is that for new features, previous deployed smart contracts won't have access to them. It is maybe okay if we consider that those features was not in the scope of the Dapps developers at the first place but it becomes a problem when we modify existing features (with the need to add or remove parameters for example) or when those features are more "user" centrics and we deprives old contract from new "cool" features that we implement.

But this is a debatable topic.